File: 0x00 基本算法/0x02 递推与递归/AcWing96.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 15;
int d[N], f[N]; // d[n]/f[n]分别表示求解该n盘3/4塔问题的最少步数

int main() {
  d[0] = 0;
  for (int i = 1; i <= 12; i ++ ) d[i] = d[i - 1] + 1 + d[i - 1];

  memset(f, 0x3f, sizeof f);
  f[0] = 0;
  for (int i = 1; i <= 12; i ++ )
    for (int j = 0; j < i; j ++ )
      f[i] = min(f[i], f[j] + d[i - j] + f[j]);

  for (int i = 1; i <= 12; i ++ ) cout << f[i] << endl;
  return 0;
}

File: 0x00 基本算法/0x02 递推与递归/AcWing96.py
def main():
    d = [2**i - 1 for i in range(13)]
    f = [0] * 13
    
    for i in range(1, 13):
        f[i] = min(2 * f[j] + d[i - j] for j in range(i))
        print(f[i])

if __name__ == '__main__':
    main()